### PR TITLE
WIP: First draft of alpha release criteria for review

### LINKED ISSUE #1362: Create a Go/No-Go Criteria Checklist for the first alpha
Let's create a list of criteria for Zebra's first alpha release.

Our developer documentation is stored here:
https://github.com/ZcashFoundation/zebra/tree/main/book/src/dev

And linked from the summary doc at:
https://github.com/ZcashFoundation/zebra/blob/main/book/src/SUMMARY.md

### PR DESCRIPTION
**Pull request by mpguerra**
[Rendered](https://github.com/ZcashFoundation/zebra/blob/pili-release-criteria/book/src/dev/release-criteria.md)

## Motivation

We need some way to help us make a decision about whether we want to go ahead with our Alpha release on the planned date.

## Solution

In order to make this decision, it is helpful to have a Go/No-Go checklist with a list of criteria that should be met in order for us to be confident in delivering our first Alpha release.

This checklist should only include measures which are **critical** for the successful launch of our first Alpha release.

For each of these measures we will assess the release status by selecting one of the following markers:
    - Green: all work is complete, with no quality issues outstanding.
    - Amber: work is not complete, but there is a revised delivery date in place and work is on track for release date.
    - Red: this work is not complete. There is no revised delivery date, or the work is off track and likely to impact release date.

- If any item on the list is red, we should delay the release of our Alpha.
- If every item is green, we can agree to release as planned.
- Amber items are up for discussion, and the decision to release should be discussed by the team.

## Review

@ZcashFoundation/zebra-team

These items are not set in stone and we will likely want to add and/or remove some as not critical to the successful Alpha release. I wanted to keep all items that were discussed in for future go/no-go checklists.

I will need help filling out _TBD_ items, which we should document and may belong elsewhere in a separate document.

## Related Issues

closes #1362

## Follow Up Work

- Release checklist
- Update for subsequent releases

### CODE REVIEW

File: book/src/dev/release-criteria.md
Comment (teor2345) [inline book/src/dev/release-criteria.md]: @dconnolly can you copy the CI configs into this suggestion?

```suggestion
Our CI tests that Zebra builds, passes its tests, and syncs to tip on the following configurations:
* Ubuntu Linux ...

Our CI tests that Zebra builds and passes its tests on the following configurations:
* recent macOS ...
* recent Windows ...
```

File: book/src/dev/release-criteria.md
Comment (teor2345) [inline book/src/dev/release-criteria.md]: @dconnolly please review, I don't know if we want to get more specific

```suggestion
Our CI builds Zebra with:
* Rust stable
* recent llvm
* recent clang
* recent libclang
```

File: book/src/dev/release-criteria.md
Comment (teor2345) [inline book/src/dev/release-criteria.md]:
```suggestion
While Zebra is still in alpha, we don't guarantee support for any particular platforms.

But you'll probably get the best results with a recent Ubuntu Linux, or the other platforms that our CI runs on.
```

File: book/src/dev/release-criteria.md
Comment (teor2345) [inline book/src/dev/release-criteria.md:25]: Let's reduce the number of checklist items here, and the level of quality we're guaranteeing?

```suggestion
        - [ ] under excellent network conditions (within 2 - 5 hours)
        - reasonable and sub-optimal network conditions are not yet supported
```

Comment (teor2345) [inline book/src/dev/release-criteria.md:25]: Let's reduce the number of checklist items here, and the level of quality we're guaranteeing?

```suggestion
    - [ ] zebrad can sync to mainnet tip
        - [ ] under excellent network conditions (within 2 - 5 hours)
        - reasonable and sub-optimal network conditions are not yet supported
```

File: book/src/dev/release-criteria.md
Comment (teor2345) [inline book/src/dev/release-criteria.md]: Let's reduce the number of checklist items here, and the level of quality we're guaranteeing?

```suggestion
    - [ ] zebrad can sync to mainnet tip
        - [ ] under excellent network conditions (within 2 - 5 hours)
        - reasonable and sub-optimal network conditions are not yet supported
    - [ ] zebrad can stay within a few blocks of the mainnet tip after the initial sync
        - [ ] under excellent network conditions
        - reasonable and sub-optimal network conditions are not yet supported
```

File: book/src/dev/release-criteria.md
Comment (teor2345) [inline book/src/dev/release-criteria.md]:
```suggestion
    - [ ] zebrad functionality works on platforms that meet its system requirements
```

File: book/src/dev/release-criteria.md
Comment (teor2345) [inline book/src/dev/release-criteria.md]: Let's remove network readiness for now, or move it to a future releases section.

- Interacting with other nodes is a duplicate of "zebrad can serve blocks to peers"
- we don't need a mechanism to deal with misbehaving nodes until the first stable release

```suggestion
```

File: book/src/dev/release-criteria.md
Comment (teor2345) [inline book/src/dev/release-criteria.md]:
```suggestion
    - [ ] Build completes within 20 minutes in Zebra's CI
```

File: book/src/dev/release-criteria.md
Comment (teor2345) [inline book/src/dev/release-criteria.md]:
```suggestion
    - [ ] All release blocker bugs have been fixed
```

File: book/src/dev/release-criteria.md
Comment (teor2345) [inline book/src/dev/release-criteria.md]: Are release notes, known issues documentation, and deployment documentation essential for the first alpha?

File: book/src/dev/release-criteria.md
Comment (teor2345) [inline book/src/dev/release-criteria.md]: I don't think we have time to finish all the outstanding troubleshooting and issue reporting features before the first alpha.

Let's move these checklist items to a future releases section.

File: book/src/dev/release-criteria.md
Comment (teor2345) [inline book/src/dev/release-criteria.md]: I don't think we can commit to this level of quality for the first alpha release
```suggestion
    - [ ] zebrad executes normally
        - [ ] zebrad's default logging works reasonably well in a terminal
        - [ ] panics, error logs, and warning logs are rare
        - [ ] known panics, errors and warnings have open tickets
```

**Review by teor2345 — CHANGES_REQUESTED** [review]
This list is a good start, but I'd like us to distinguish between:
* essential things for the first alpha release
* things that can wait for future releases

I also think we need to focus our efforts on things that are achievable in the next week.

I have a bunch of suggestions, but I'm missing some info that I think @dconnolly can help fill in.

I also think we should be consistent with the `zebrad` command name, and spell it `zebrad` (all lower case), even when it's at the start of a sentence. That way, we match the command-line spelling.

File: book/src/dev/release-criteria.md
Comment (mpguerra) [reply book/src/dev/release-criteria.md]: I think known issues should be included at least so users know what to expect. Happy to discuss further

File: book/src/dev/release-criteria.md
Comment (yaahc) [inline book/src/dev/release-criteria.md]: how are we going to track the color status on these items? My guess would be emojis. And are we going to do it in this file or out of band somehow?

:red_circle: :yellow_circle: :green_circle:

File: book/src/dev/release-criteria.md
Comment (yaahc) [inline book/src/dev/release-criteria.md]: My intution is that it would be better to enumerate the bugs we know of rather than have a catch all checklist item.

File: book/src/dev/release-criteria.md
Comment (yaahc) [inline book/src/dev/release-criteria.md]: how do these check boxes render in `mdbook`? I am worried this might be a github specific feature that wont translate to our book.

File: book/src/dev/release-criteria.md
Comment (yaahc) [inline book/src/dev/release-criteria.md]: These feel vague / hard to measure

File: book/src/dev/release-criteria.md
Comment (yaahc) [inline book/src/dev/release-criteria.md]: was this section accidentally duplicated?

File: book/src/dev/release-criteria.md
Comment (mpguerra) [reply book/src/dev/release-criteria.md]: Good questions, emojis are good. 
We should set aside some time (5-10 minutes) during the next Zebra Gardening calls (on 11/30 and 12/07) to go through the checklist and update the status (RAG) for each item by replacing the checkboxes with the appropriate emoji.

File: book/src/dev/release-criteria.md
Comment (mpguerra) [reply book/src/dev/release-criteria.md]: let's use the emoji instead when we do the first update to the status of each item.

File: book/src/dev/release-criteria.md
Comment (mpguerra) [reply book/src/dev/release-criteria.md]: I agree, let's discuss either during the zebra gardening call, a separate meeting, or we can try an async triage style process like we did for the tickets needing triage.

File: book/src/dev/release-criteria.md
Comment (mpguerra) [reply book/src/dev/release-criteria.md]: This is for future releases since I trimmed the original broader list after @teor2345 's review and I didn't want to lose the details.

File: book/src/dev/release-criteria.md
Comment (mpguerra) [reply book/src/dev/release-criteria.md]: true, let's all think about how we could measure this and make some suggestions

File: book/src/dev/release-criteria.md
Comment (teor2345) [reply book/src/dev/release-criteria.md]: Sounds good - but since many users will just build the `main` branch, the list of known issues will change over time.

So let's include a link to a GitHub query - something like "severe bug" or "release bug".

We don't rate the severity of our bugs yet. But we can't just send people to the full list of bugs, because it contains a lot of bugs they probably won't care about.

File: book/src/dev/release-criteria.md
Comment (teor2345) [reply book/src/dev/release-criteria.md]: We could take a lot of time talking about the right way to do logging or errors. So let's try to limit how much effort we put into reaching a conclusion here?

I'd rather spend that time on bugs that we all agree are bugs - and that affect the user experience more than logging, panics, errors, or warnings.

File: book/src/dev/release-criteria.md
Comment (teor2345) [reply book/src/dev/release-criteria.md]: I've un-resolved this conversation, because @dconnolly hasn't filled in this info yet.

File: book/src/dev/release-criteria.md
Comment (teor2345) [reply book/src/dev/release-criteria.md]: I think we should have a GitHub tag for release blocker bugs, rather than repeating that information in another document.

File: book/src/dev/release-criteria.md
Comment (teor2345) [reply book/src/dev/release-criteria.md]: Can you put this duplicate info in a separate file, ticket, or pull request?

It's not clear which parts of the document apply to which releases.

And in general, it's going to be harder to maintain two lists that are partial duplicates of each other. Maybe we should keep a file with the extra items that aren't included in the final list?

**Review by teor2345 — COMMENTED** [review]
I'd like us to put the future release list in a separate file - or in a follow-up ticket.

Having duplicate info in the same document can be confusing.

File: book/src/dev/release-criteria.md
Comment (mpguerra) [reply book/src/dev/release-criteria.md]: I've added a link to the existing issues labelled with "C-bug" that are on the alpha release milestone. We should definitely add some additional priority and severity labels to our repo after alpha to help with triaging external bug reports and help us define what constitutes a blocking bug for future releases.

File: book/src/dev/release-criteria.md
Comment (mpguerra) [reply book/src/dev/release-criteria.md]: +1

File: book/src/dev/release-criteria.md
Comment (mpguerra) [reply book/src/dev/release-criteria.md]: done

File: book/src/dev/release-criteria.md
Comment (teor2345) [inline book/src/dev/release-criteria.md]: Until PR #1444 merges, this feature is incomplete
```suggestion
    - ⚠️  `zebrad` can validate proof of work
```

File: book/src/dev/release-criteria.md
Comment (teor2345) [inline book/src/dev/release-criteria.md]: Most builds seem to complete within 30 minutes, but I've seen the occasional slow build.
Some of our build steps seem to be re-compiling things they don't need to.
```suggestion
    - ⚠️ Build completes within 30 minutes in Zebra's CI
```

File: book/src/dev/release-criteria.md
Comment (teor2345) [inline book/src/dev/release-criteria.md]: I think we're doing pretty well here right now
```suggestion
    - ✅ `zebrad` executes normally
        - ✅ `zebrad`'s default logging works reasonably well in a terminal
        - ✅ panics, error logs, and warning logs are rare on mainnet
        - ✅ known panics, errors and warnings have open tickets
```

**Review by teor2345 — DISMISSED** [review]
This is looking really good.

I've suggested a few changes based on recent progress.

We haven't added any system requirements, I've added that task to #1419.

File: book/src/dev/release-criteria.md
Comment (mpguerra) [reply book/src/dev/release-criteria.md]: I agree it can be hard to measure these more qualitative criteria. I'll leave them as is for now and we can refine for future releases.

**Comment by mpguerra** [comment]
I'm going to merge this for now so we have the file in main and create a new ticket to update system requirements and assign to @dconnolly

**Pull request merged**
